katz/src/main/kotlin/katz/Either.kt

```diff
@@ -0,0 +1,196 @@
+
+    /**
+     * Returns a [Option.Some] containing the [Right] value
+     * if it exists or a [Option.None] if this is a [Left].
+     *
+     * Example:
+     * ```
+     * Right(12).toOption() // Result: Some(12)
+     * Left(12).toOption()  // Result: None
+     * ```
+     */
+    fun toOption(): Option<B> =
+            fold({ Option.None }, { Option.Some(it) })
+
+    /**
+     * The left side of the disjoint union, as opposed to the [Right] side.
+     */
+    data class Left<out A : Any>(val a: A) : Either<A, Nothing>() {
+        override val isLeft = true
+        override val isRight = false
+    }
+
+    /**
+     * The right side of the disjoint union, as opposed to the [Left] side.
+     */
+    data class Right<out B : Any>(val b: B) : Either<Nothing, B>() {
+        override val isLeft = false
+        override val isRight = true
+    }
+}
+
+/**
+ * Binds the given function across [Either.Right].
+ *
+ * @param f The function to bind across [Either.Right].
+ */
+inline fun <A : Any, B : Any, C : Any> Either<A, B>.flatMap(f: (B) -> Either<A, C>): Either<A, C> =
+        fold({ Either.Left(it) }, { f(it) })
+
+/**
+ * Returns the value from this [Either.Right] or the given argument if this is a [Either.Left].
+ *
+ * Example:
+ * ```
+ * Right(12).getOrElse(17) // Result: 12
+ * Left(12).getOrElse(17)  // Result: 17
+ * ```
+ */
+inline fun <B : Any> Either<*, B>.getOrElse(default: () -> B): B =
+        fold({ default() }, { it })
+
+/**
+ * * Returns [Either.Right] with the existing value of [Either.Right] if this is a [Either.Right] and the given predicate
+ * holds for the right value.
+ * * Returns `Left(default)` if this is a [Either.Right] and the given predicate does not
+ * hold for the right value.
+ * * Returns [Either.Left] with the existing value of [Either.Left] if this is a [Either.Left].
+ *
+ * Example:
+ * ```
+ * Right(12).filterOrElse({ it > 10 }, { -1 }) // Result: Right(12)
+ * Right(7).filterOrElse({ it > 10 }, { -1 })  // Result: Left(-1)
+ *
+ * val left: Either<Int, Int> = Left(12)
+ * left.filterOrElse({ it > 10 }, { -1 })      // Result: Left(12)
+ * ```
+ */
+inline fun <A : Any, B : Any> Either<A, B>.filterOrElse(predicate: (B) -> Boolean, default: () -> A): Either<A, B> =
+        fold({ Either.Left(it) }, { if (predicate(it)) Either.Right(it) else Either.Left(default()) })
+
+/**
+ * Returns `true` if this is a [Right] and its value is equal to `elem` (as determined by `==`),
+ * returns `false` otherwise.
+ *
+ * Example:
+ * ```
+ * Right("something").contains { "something" } // Result: true
+ * Right("something").contains { "anything" }  // Result: false
+ * Left("something").contains { "something" }  // Result: false
+ *  ```
+ *
+ * @param elem    the element to test.
+ * @return `true` if the option has an element that is equal (as determined by `==`) to `elem`, `false` otherwise.
+ */
+fun <A : Any, B, BB : Any> Either<A, B>.contains(elem: () -> BB): Boolean where B : BB =
```

Comment:
`elem` should be a simple type not a function, this is the way:

`fun <A : Any, B : Any> Either<A, B>.contains(elem: B): Boolean = fold({ false }, { it == elem })`